2048

matsarat opened the pull request.

matsarat committed:
Initial commit to 2048 game

Code generating board and two initial fields with value of "2"

matsarat committed:
Add Field class, moving field values to the right and left

matsarat committed:
Add move_up and move_down methods

matsarat committed:
Add summing values in rows to the right

matsarat committed:
Add summing values in four directions (up, down, left, right)

matsarat committed:
Place next element on the board

matsarat committed:
Add function that ends the game.

matsarat committed:
Add function that validates player's move.

matsarat committed:
Add game functionality, needs correction in summation during move

matsarat committed:
Correct placing two initial elements on the board

matsarat committed:
Correct in game_over method

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,223 @@
+    board = create_board(number_of_rows, number_of_columns)
+    instructions()
+    place_two_initial_elements_on_board(board)
+    print_board(board)
```
you don't need this method here

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,223 @@
+    print_board(board)
+    while game_over(board) == False:
+        valid_players_move(board)
+        get_fields_with_zero_value(board)
```
you don't need this method here

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,223 @@
+    while game_over(board) == False:
+        valid_players_move(board)
+        get_fields_with_zero_value(board)
+        game_over(board)
```
you don't need this method here

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,223 @@
+            return True
+    if len(get_fields_with_zero_value(board)) > 0:
+        return False
+    if len(get_fields_with_zero_value(board)) == 0:
```
you don't need this if here

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,223 @@
+        if winning_number in row:
+            print("You did it!")
+            return True
+    if len(get_fields_with_zero_value(board)) > 0:
```
you should add here `or` when we have on board situation where two fields have same value

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,223 @@
+    print("To move up - press 'w'")
+
+
+def place_two_initial_elements_on_board(board):
```
use `def get_fields_with_zero_value(board)` here

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,223 @@
+    return fields_with_zero_value
+
+def place_next_element_on_board(board):
+    next_field = random.choice(get_fields_with_zero_value(board))
```
`random.choice()` docs:
```doc
Return a random element from the non-empty sequence seq. If seq is empty, raises IndexError.
```
IMHO with current logic this method can raise `IndexError`

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,223 @@
+    if players_move == "a":
+        move_left(board)
+        sum_left(board)
+        move_left(board)
```
hmm... If I remember our design of this method it didn't look like this implementation :D Check your notes!

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,223 @@
+
+
+def place_two_initial_elements_on_board(board):
+    initial_element_value = 2
+    get_fields_with_zero_value(board)
+
+    first_initial_element = random.choice(get_fields_with_zero_value(board))
+    first_initial_element.value = initial_element_value
+
+    get_fields_with_zero_value(board)
+    second_initial_element = random.choice(get_fields_with_zero_value(board))
+    second_initial_element.value = initial_element_value
```
use `def get_fields_with_zero_value(board)` here

matsarat committed:
Correct in game method

matsarat committed:
Correct summation during player's move

matsarat committed:
Correct summation during player's move

matsarat committed:
Correct ending game conditions

Added function that checks if on the board are two similar neighboring elements

matsarat committed:
Correct ending game conditions

Added function that checks if on the board exists winning number

matsarat committed:
Apply printing board after winning the game

matsarat committed:
Apply validation of player's move and placing new element on the board

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,275 @@
+    return False
+
+
+def get_temporary_board_values(board):
```
What is the difference between these two methods: `get_temporary_board_values` and `get_board_values`?

matsarat replied:
One of these got deleted and correction in method previously using both of them was made.

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,275 @@
+        sum_left(board)
+        move_left(board)
+        values_board = get_board_values(board)
+        if temporary_values_board == values_board:
```
this sequence is repeated 4 times!!! What can you do with it? :D

matsarat replied:
Method was extracted and called check_if_move_is_valid .

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,275 @@
+    instructions()
+    place_two_initial_elements_on_board(board)
+    print_board(board)
+    while game_over(board) == False or check_if_similar_neighboring(board) == True:
```
invoke `check_if_similar_neighboring(board)` inside `gamve_over` method. Also you can have: `while not game_over(board)` which is easier to read

matsarat replied:
Done.

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,275 @@
+
+
+def check_if_won(board):
+    winning_number = 16
```
hmm :D 16? :D

matsarat replied:
Convenient during tests. Corrected.

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,275 @@
+            break
+        print_board(board)
+
+game(3, 3)
```
4,4

matsarat replied:
Convenient during tests. Corrected.

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,275 @@
+    board = create_board(number_of_rows, number_of_columns)
+    instructions()
+    place_two_initial_elements_on_board(board)
+    print_board(board)
```
with current logic this print for sure is not necessary! You can `print_board` firstly in `while` loop

matsarat replied:
Corrected. Also added printing board after losing game in game_over method.

matsarat committed:
Clean up code and correct conditions - board size and winning number.

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,265 @@
+    while not game_over(board):
+        print_board(board)
+        validated_players_move(board)
+        if check_if_won(board) == True:
```
FYI - this method also can be moved to `game_over` method (minor - you do not have to do that)

Czeffik commented on 2048.py:
```diff
@@ -0,0 +1,265 @@
+    place_two_initial_elements_on_board(board)
+    while not game_over(board):
+        print_board(board)
+        validated_players_move(board)
```
FYI - for me it would be clearer when this method would be named `make_move` (minor - you do not have to do that)
